### PR TITLE
fix bug 982 : durbin watson for several inputs

### PR DESCRIPTION
**Pull request by adumasphi**
I forgot to take into account several factors when building the regression matrix.
I noticed also an error when computing the p-value, that I corrected.
A validation python file, comparing result with OT and R is provided in validation/src/ValidDurbinWatson.py

As an improvement, it could be nice to have the statistic value in the result object.

### CODE REVIEW

**Comment by sofianehaddad** [comment]
@adumasphi thanks for your feedback. I will have a look at the PR
I agreed concerning the statistic in TestResult. Maybe in ot 1.13...

File: lib/src/Uncertainty/StatTests/LinearModelTest.cxx [hunk 1]
```diff
@@ -287,19 +298,23 @@ TestResult LinearModelTest::LinearModelDurbinWatson(const Sample & firstSample,
...
-  Scalar pValue = 2.0 * DistFunc::pNormal(std::abs(dw - dmean) / std::sqrt(dvar), true);
-  Description description(1, "Hypothesis test: autocorrelation equals 0.");
+  Scalar pValue;
+  Description description(1, "");
+  if (hypothesis == "Equal")
+  {
+    pValue = 2.0 * DistFunc::pNormal(std::abs(dw - dmean) / std::sqrt(dvar), true);
+    description[0] = "Hypothesis test: autocorrelation equals 0.";
+  }
   if(hypothesis == "Less")
```
Comment (jschueller) [inline lib/src/Uncertainty/StatTests/LinearModelTest.cxx]: else if

File: lib/src/Uncertainty/StatTests/LinearModelTest.cxx [hunk 2]
```diff
@@ -287,19 +298,23 @@ TestResult LinearModelTest::LinearModelDurbinWatson(const Sample & firstSample,
...
-  Scalar pValue = 2.0 * DistFunc::pNormal(std::abs(dw - dmean) / std::sqrt(dvar), true);
-  Description description(1, "Hypothesis test: autocorrelation equals 0.");
-  if(hypothesis == "Less")
+  Scalar pValue;
+  Description description(1, "");
+  if (hypothesis == "Equal")
+  {
+    pValue = 2.0 * DistFunc::pNormal(std::abs(dw - dmean) / std::sqrt(dvar), true);
+    description[0] = "Hypothesis test: autocorrelation equals 0.";
+  }
```
Comment (sofianehaddad) [inline lib/src/Uncertainty/StatTests/LinearModelTest.cxx]: maybe shorter description ? 

```
H0: auto.cor=0
```

File: lib/src/Uncertainty/StatTests/LinearModelTest.cxx [hunk 3]
```diff
@@ -287,19 +298,23 @@ TestResult LinearModelTest::LinearModelDurbinWatson(const Sample & firstSample,
...
+  else if(hypothesis == "Less")
   {
-    pValue = 1 - pValue / 2;
+    pValue = DistFunc::pNormal((dw - dmean) / std::sqrt(dvar));
     description[0] = "Hypothesis test: autocorrelation is less than 0";
   }
```
Comment (sofianehaddad) [inline lib/src/Uncertainty/StatTests/LinearModelTest.cxx]: same here for description

File: lib/src/Uncertainty/StatTests/LinearModelTest.cxx [hunk 4]
```diff
@@ -287,19 +298,23 @@ TestResult LinearModelTest::LinearModelDurbinWatson(const Sample & firstSample,
...
   else if(hypothesis == "Greater")
   {
-    pValue = pValue / 2;
+    pValue = DistFunc::pNormal((dw - dmean) / std::sqrt(dvar), true);
     description[0] = "Hypothesis test: autocorrelation is greater than 0";
   }
```
Comment (sofianehaddad) [inline lib/src/Uncertainty/StatTests/LinearModelTest.cxx]: same here for description

**Comment by sofianehaddad** [comment]
@adumasphi could you please add the output validation file ? it helps comparing with potential future changes in API/R

File: lib/src/Uncertainty/StatTests/LinearModelTest.cxx [hunk 5]
```diff
@@ -266,18 +266,29 @@ TestResult LinearModelTest::LinearModelDurbinWatson(const Sample & firstSample,
...
-    X(i, 0) = 1;
-    X(i, 1) = firstSample(i, 0);
+    for(UnsignedInteger i = 0; i < residualSize; ++i)
+    {
+      if (j == 0)
+        X(i, j) = 1.0;
+      else
+        X(i, j) = firstSample(i, j-1);
+    }
   }
```
[the remark this answers is not in the document]
Comment (adumasphi) [reply lib/src/Uncertainty/StatTests/LinearModelTest.cxx]: ok agreed

File: lib/src/Uncertainty/StatTests/LinearModelTest.cxx (hunk 2 above)
Comment (adumasphi) [reply lib/src/Uncertainty/StatTests/LinearModelTest.cxx]: Ok it's ok for me.

**Comment by jschueller** [comment]
Antoine, the docstring test still fails.

File: lib/src/Uncertainty/StatTests/LinearModelTest.cxx [hunk 6]
```diff
@@ -287,20 +296,24 @@ TestResult LinearModelTest::LinearModelDurbinWatson(const Sample & firstSample,
   const Scalar Q = 2 * (3 * residualSize - 4) - 2 * XAXTrace + (XAXQt * XAXQt).getImplementation()->computeTrace();
   const Scalar dmean = P / (residualSize - (dimension + 1));
   const Scalar dvar = 2.0 / ((residualSize - (dimension + 1)) * (residualSize - (dimension + 1) + 2)) * (Q - P * dmean);
-
+
```
Comment (regislebrun) [inline lib/src/Uncertainty/StatTests/LinearModelTest.cxx]: spurious space

File: lib/src/Uncertainty/StatTests/LinearModelTest.cxx [hunk 7]
```diff
@@ -287,20 +296,24 @@ TestResult LinearModelTest::LinearModelDurbinWatson(const Sample & firstSample,
...
   const Scalar dvar = 2.0 / ((residualSize - (dimension + 1)) * (residualSize - (dimension + 1) + 2)) * (Q - P * dmean);
-
+  
   /* Compute the p-value with respect to the hypothesis */
-  // Initial values defined for hypothesis = "Equal"
-  Scalar pValue = 2.0 * DistFunc::pNormal(std::abs(dw - dmean) / std::sqrt(dvar), true);
-  Description description(1, "Hypothesis test: autocorrelation equals 0.");
-  if(hypothesis == "Less")
+  Scalar pValue;
+  Description description(1, "");
```
Comment (regislebrun) [inline lib/src/Uncertainty/StatTests/LinearModelTest.cxx]: or simply description(1)

Comment (adumasphi) [reply lib/src/Uncertainty/StatTests/LinearModelTest.cxx]: ok changes are done.

**Pull request merged**